pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,265 @@
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing CRD")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueFromAPIBinding(obj interface{}) {
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), binding)
+
+	for _, boundResource := range binding.Status.BoundResources {
+		key := kcpcache.ToClusterAwareKey(apibinding.ShadowWorkspaceName.String(), "", boundResource.Schema.UID)
+		logging.WithQueueKey(logger, key).V(2).Info("queueing CRD via APIBinding")
+		c.queue.Add(key)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.crdLister.Cluster(clusterName).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	logger := logging.WithObject(klog.FromContext(ctx), obj)
+	ctx = klog.NewContext(ctx, logger)
+
+	result, err := c.apiBindingIndexer.ByIndex(indexers.APIBindingByBoundResourceUID, obj.Name)
+
+	if err != nil {
+		return err
+	}
+
+	bindingList, err := c.apiBindingLister.List(labels.Everything())
+
+	if err != nil {
+		return err
+	}
+
+	for _, binding := range bindingList {
+		if len(binding.Status.BoundResources) == 0 {
+			logger.V(1).Info("APIBinding has no bound resources.  Need to wait until all bindings have bound resources")
+			c.queue.AddAfter(key, time.Second)
```

Comment:
why requeue? This happens automatically. You are watch bindings.

Reply:
Ah, true.

Reply:
Turns out we do in fact need this explicit requeue.  If the `APIBinding` that was discovered to have no bound resources is not actually related to this CRD (we have no reasonable way to know), then we lost perhaps our last opportunity to delete this particular CRD (in the event that it was the deletion event for the last `APIBinding` for this CRD).

Reply:
Don't this is true. There is some case when BoundResources gets empty (could be in theory). The solution with requeue is not correct. It causes a potential DoS.

What we could do:
1. depend on resync to eventually relist the CRD
2. do something that normally a big nogo: enqueue both old and new object on update.

Reply:
The `APIBinding` controller will create the bound CRD, wait for it to become established, and _then_ it will add the `BoundResource` to the API binding's status.  I've seen the potential race condition cause a problem when running e2e tests with `-count 20`, but not `-count 5`.  

> depend on resync to eventually relist the CRD

This seems reasonable for now at least, but how can we run the e2e test if we rely on the resync to catch edge cases?  Do we need to modify or remove our new check?  Can we change the resync period for the test?

> It causes a potential DoS

Could we call `queue.AddRateLimited(key)` instead if we're only concerned about DoS?

Reply:
> The APIBinding controller will create the bound CRD, wait for it to become established, and then it will add the BoundResource to the API binding's status. I've seen the potential race condition cause a problem when running e2e tests with -count 20, but not -count 5.

Got it. Yes, this is true. Very good insight!

I see how your requeue helps. I am worried that it endlessly requeues. Thinking. Not sure how `AddRateLimited` really works internally. It's won't keep a giant per-key map with requeues, will it?

Reply:
Just thinking out loud: could we check for APIResourceSchema existence, and everything else before the actual CRD creation, and only do the latter AFTER setting the BoundResources ?

Wondering whether we run into any other race that way.

Reply:
Also the `creationTime` idea would help. Let a CRD live up to a minute or so and only delete then.

Reply:
> It's won't keep a giant per-key map with requeues, will it?
Yes, I'm afraid if some `APIBinding` somehow never made it to the `Bound` phase yet also had a valid `APIExport`, then eventually every CRD key would end up in the requeue map.  This should be considered a bug, and we could even alert if the requeue count increased too quickly for the CRD cleanup controller.

> could we check for APIResourceSchema existence, and everything else before the actual CRD creation, and only do the latter AFTER setting the BoundResources ?
Possibly, but this seems unintuitive to me.  The resource isn't actually bound until the CRD is created!

> creationTime idea would help.

If an `APIBinding` (that has a valid `APIExport`) never makes it to the `Bound` phase, then the `creationTime` idea would only delay the inevitable addition of the CRD to the requeue map.

Barring that edge case (which I believe doesn't exist today and if it did would be considered a bug), I can't think of any other scenarios that aren't covered by the code as-is, but I wouldn't mind adding a `creationTime` check as a "belts and suspenders" kind of thing.

Reply:
> then the creationTime idea would only delay the inevitable addition of the CRD to the requeue map.

No, I mean to delete the CRD after a minute, so give a minute of time to get over the race.